### PR TITLE
feat: penpal v5

### PR DESCRIPTION
**Pull request by simboonlong**
### bg context
- need to upgrade to penpal v5
- if there is a mismatch of parent child connection versions, promise will not resolve and always be pending

### attempted solution
- ~~try connect v5 connect first. if timeout, try v4 connect instead (for older template renderers out there)~~
- promise.any v5 and v4 connections, first to resolved will connect. otherwise timeout to reject connection

Screenshot 2021-03-02 at 2 42 39 PM

### CODE REVIEW

File: src/components/frame/useFrame.ts [hunk 1]
```diff
@@ -1,35 +1,58 @@
 import { RefObject, useEffect, useState } from "react";
-import connectToParent from "penpal/lib/connectToParent";
-import connectToChild from "penpal/lib/connectToChild";
+import { connectToParent, connectToChild } from "penpal";
+// eslint-disable-next-line
+// @ts-ignore
```
Comment (Nebulis) [inline src/components/frame/useFrame.ts:4]: what's the issue here ?

File: src/components/frame/useFrame.ts [hunk 2]
```diff
@@ -1,35 +1,58 @@
...
-        await connectToParent({
-          methods: {
-            dispatch: dispatch
-          }
-        }).promise
-      );
-      setStatus("CONNECTED");
-    };
     if (inIframe() && status === "DISCONNECTED") {
+      const parentV5 = connectToParent({
```
Comment (Nebulis) [inline src/components/frame/useFrame.ts:30]: I would suggest for the parent connection to use only one version, v5. the reason we have control over Parent (verifier) and we can upgrade them to support v5/v4 before upgrading the renderers.

as long as verifiers, supports v5 and v4 we can opt to one of the version. In the event one need a connection with penpal v4, can use a specific version (maybe would be nice to add in readme). Also if we support v5 in verifiers, there is no good reason to allow renderer to use v4

File: src/components/frame/useFrame.ts (hunk 1 above)
Comment (simboonlong) [reply src/components/frame/useFrame.ts]: ok fixed by updating the type declaration to the alias one

File: src/components/frame/useFrame.ts (hunk 2 above)
Comment (simboonlong) [reply src/components/frame/useFrame.ts:26]: looks like its still working with parent at only v5, i added an v4 example to play with:

Screenshot 2021-02-26 at 1 39 02 PM

File: example/application/container.tsx [hunk 3]
```diff
@@ -3,11 +3,16 @@ import { FrameActions, FrameConnector, HostActionsHandler } from "../../src";
 import { css } from "@emotion/core";
```
Comment (Nebulis) [inline example/application/container.tsx:3]: you should rebase, if not done already, Raymond did changes on this

Comment (simboonlong) [reply example/application/container.tsx:3]: yeah i rebased to latest, b4 making my changes

File: README.md [hunk 4]
```diff
@@ -190,12 +190,16 @@ Each configured `Template` will be provided the following properties, when rende
...
+
 - `npm run storybook`: to start storybook, create stories and visualize the different component
 - `npm run test`: to run tests
 - `npm run lint`: to run lint
 - `npm run example:application`: to run the example build with the library to develop an hosting application. Don't forget to update the example if you update this library.
 - `npm run example:renderer`: to run the example build with the library to develop a decentralized renderer. Don't forget to update the example if you update this library.
 
 We also provided a [github template](https://github.com/Open-Attestation/decentralized-renderer-react-template) to build your own decentralized-renderer based on this library
+
+## Penpal
```
Comment (Nebulis) [inline README.md:203]: not true right ? We try to connect with both version

Comment (Nebulis) [reply README.md:203]: actually true bur for renderer only

File: src/components/frame/useFrame.ts [hunk 5]
```diff
@@ -43,26 +53,42 @@ interface UseChildrenFrameProps {
...
-      setStatus("CONNECTED");
-    };
     if (props.iframe.current && status === "DISCONNECTED") {
+      const childV5 = connectToChild({
+        methods: {
+          dispatch: props.dispatch,
+          ...props.methods,
+        },
+        iframe: props.iframe.current,
+        timeout: 5000, // this will ensure connection to promise reject, when connection versions not match. otherwise it will be stucked in promise pending
```
Comment (Nebulis) [inline src/components/frame/useFrame.ts:67]: I'm scared that 5s timeout is a bit low, for instance cold start of server + react :grimacing:

Comment (simboonlong) [reply src/components/frame/useFrame.ts]: ok 9000 then

Comment (Nebulis) [reply src/components/frame/useFrame.ts]: why 9 000 ?

Comment (simboonlong) [reply src/components/frame/useFrame.ts]: no particular reason, if 5 digits the finger need to apply extra keystroke

Comment (Nebulis) [reply src/components/frame/useFrame.ts]: Lol let's go with 30s instead (same as browser default timeout for some stuff)

Comment (simboonlong) [reply src/components/frame/useFrame.ts]: ok no prob

**Review by Nebulis — APPROVED** [review]

**Pull request merged**